model/lifecycle.go

```diff
@@ -432,79 +450,18 @@ func RestartVersions(versionsToRestart []*VersionToRestart, abortInProgress bool
 func RestartBuild(buildId string, taskIds []string, abortInProgress bool, caller string) error {
 	if abortInProgress {
 		// abort in-progress tasks in this build
-		if err := task.AbortTasksForBuild(buildId, taskIds, caller); err != nil {
+		if err := task.AbortAndMarkResetTasksForBuild(buildId, taskIds, caller); err != nil {
 			return errors.WithStack(err)
 		}
 	}
-
-	// restart all the 'not in-progress' tasks for the build
-	tasks, err := task.FindAll(db.Query(task.ByIdsAndStatus(taskIds, evergreen.TaskCompletedStatuses)))
+	tasksToReset, err := getTasksToReset(taskIds)
 	if err != nil {
-		return errors.WithStack(err)
+		return errors.Wrap(err, "getting finished tasks")
 	}
-	if len(tasks) == 0 {
+	if len(tasksToReset) == 0 {
 		return nil
 	}
-	return restartTasksForBuild(tasks, caller)
-}
-
-// RestartAllBuildTasks restarts all the tasks associated with a given build.
-func RestartAllBuildTasks(buildId string, caller string) error {
-	if err := task.AbortTasksForBuild(buildId, nil, caller); err != nil {
-		return errors.WithStack(err)
-	}
-
-	allTasks, err := task.FindAll(db.Query(task.ByBuildId(buildId)))
-	if err != nil {
-		return errors.WithStack(err)
-	}
-	if len(allTasks) == 0 {
-		return nil
-	}
-	return restartTasksForBuild(allTasks, caller)
-}
-
-func restartTasksForBuild(tasks []task.Task, caller string) error {
-	// maps task group to a single task in the group so we only check once
-	taskGroupsToCheck := map[string]task.Task{}
-	restartIds := []string{}
-	toArchive := []task.Task{}
-	for _, t := range tasks {
-		if t.IsPartOfSingleHostTaskGroup() {
-			if err := t.SetResetWhenFinished(); err != nil {
-				return errors.Wrapf(err, "marking task group '%s' to reset", t.TaskGroup)
-			}
-			taskGroupsToCheck[t.TaskGroup] = t
-		} else {
-			restartIds = append(restartIds, t.Id)
-			if t.DisplayOnly {
-				restartIds = append(restartIds, t.ExecutionTasks...)
-			}
-			if t.IsFinished() {
-				toArchive = append(toArchive, t)
-			}
-		}
-	}
-	if err := task.ArchiveMany(toArchive); err != nil {
-		return errors.Wrap(err, "archiving tasks")
-	}
-	// Set all the task fields to indicate restarted
-	if err := MarkTasksReset(restartIds); err != nil {
-		return errors.WithStack(err)
-	}
-	for _, t := range tasks {
-		if !t.IsPartOfSingleHostTaskGroup() { // this will be logged separately if task group is restarted
-			event.LogTaskRestarted(t.Id, t.Execution, caller)
-		}
-	}
-
-	for tg, t := range taskGroupsToCheck {
-		if err := checkResetSingleHostTaskGroup(&t, caller); err != nil {
-			return errors.Wrapf(err, "resetting single host task group '%s'", tg)
-		}
-	}
-
-	return errors.Wrap(build.SetBuildStartedForTasks(tasks, caller), "setting builds started")
+	return restartTasks(tasksToReset, caller)
```

Comment:
Since etVersionStatus was moved out of restartTasks, I think that the version status stays outdated here. I'm not sure there's a time we'd want to restart tasks and _not_ update the version so we may want to move it back in.

Reply:
That is a good point. I think moving `setVersionStatus` should fix this.